src/main.rs

```diff
@@ -77,10 +81,33 @@ enum UpdateError {
     RequestError(#[from] request::RequestError),
 }
 
+async fn perform_with_delay<T>(
+    task: T,
+    previous_ts: Arc<TMutex<Instant>>,
+    delay: Duration,
+) -> T::Output
+where
+    T: Future,
+{
+    let mut locked_ts = previous_ts.lock().await;
+    loop {
+        let time_passed = Instant::now().duration_since(*locked_ts);
+
+        if time_passed >= delay {
+            let res = task.await;
+            *locked_ts = Instant::now();
+            return res;
+        } else {
+            tokio::time::sleep(delay - time_passed).await;
+        }
+    }
```

Comment:
Does this loop ever actually... loop? Would something a bit like this not work instead?
```suggestion
    let time_passed = Instant::now().duration_since(*locked_ts);

    if time_passed < delay {
        tokio::time::sleep(delay - time_passed).await;
    }

    *locked_ts = Instant::now();
    }
```
then just removing the `task` argument and using this as something to wait on?

Reply:
Yes, I also thought about removing the loop, so will do